godot-core/src/builtin/projection.rs

```diff
@@ -770,4 +781,200 @@ mod test {
+                RMat4::from_cols_array_2d(&mat),
+                matrix_eq_approx,
+                "left={left} right={right} bottom={bottom} top={top} near={near} far={far}"
+            );
+        }
+    }
+
+    #[test]
+    fn test_frustum_aspect() {
+        const TEST_DATA: [((real, real, Vector2, real, real, bool), [[real; 4]; 4]); 4] = [
+            (
+                (2.0, 1.0, Vector2::ZERO, 1.0, 2.0, false),
+                [
+                    [1.0, 0.0, 0.0, 0.0],
+                    [0.0, 1.0, 0.0, 0.0],
+                    [0.0, 0.0, -3.0, -1.0],
+                    [0.0, 0.0, -4.0, 0.0],
+                ],
+            ),
+            (
+                (2.0, 1.0, Vector2::ZERO, 1.0, 2.0, true),
+                [
+                    [1.0, 0.0, 0.0, 0.0],
+                    [0.0, 1.0, 0.0, 0.0],
+                    [0.0, 0.0, -3.0, -1.0],
+                    [0.0, 0.0, -4.0, 0.0],
+                ],
+            ),
+            (
+                (1.0, 1.0, Vector2::new(0.5, 0.5), 1.0, 2.0, false),
+                [
+                    [2.0, 0.0, 0.0, 0.0],
+                    [0.0, 2.0, 0.0, 0.0],
+                    [1.0, 1.0, -3.0, -1.0],
+                    [0.0, 0.0, -4.0, 0.0],
+                ],
+            ),
+            (
+                (0.05, 4.0, Vector2::ZERO, 0.05, 100.0, false),
+                [
+                    [0.5, 0.0, 0.0, 0.0],
+                    [0.0, 2.0, 0.0, 0.0],
+                    [0.0, 0.0, -1.001000500250125, -1.0],
+                    [0.0, 0.0, -0.10005002501250625, 0.0],
+                ],
+            ),
+        ];
+
+        for ((size, aspect, offset, near, far, flip_fov), mat) in TEST_DATA {
+            assert_eq_approx!(
+                Projection::create_frustum_aspect(size, aspect, offset, near, far, flip_fov),
+                RMat4::from_cols_array_2d(&mat),
+                matrix_eq_approx,
+                "size={size} aspect={aspect} offset=({0} {1}) near={near} far={far} flip_fov={flip_fov}",
+                offset.x,
+                offset.y,
+            );
+        }
+    }
+
+    // TODO: Test create_for_hmd, create_perspective_hmd
+
+    #[test]
+    fn test_is_orthogonal() {
+        // Orthogonal
+        for left_i in 0..20 {
+            for right_i in left_i + 1..=20 {
+                for bottom_i in 0..20 {
+                    for top_i in bottom_i + 1..=20 {
+                        for near_i in 0..20 {
+                            for far_i in near_i + 1..=20 {
+                                let left = (left_i as real) * 0.5 - 5.0;
+                                let right = (right_i as real) * 0.5 - 5.0;
+                                let bottom = (bottom_i as real) * 0.5 - 5.0;
+                                let top = (top_i as real) * 0.5 - 5.0;
+                                let near = (near_i as real) * 0.5 - 5.0;
+                                let far = (far_i as real) * 0.5 - 5.0;
```

Comment:
Maybe you could extract the `(x as real) * 0.5 - 5.0` into a small closure or function.

Reply:
Probably, but better solution would be proper fuzzing. I'll refactor it.